Fix changing column from int to bigint with autoincrement

icewind1991 opened the pull request:
Currently it tries to alter the column type to `BIGSERIAL` which is not a valid type to be used in an `ALTER TABLE` query.

Also it currently drops the default which breaks the auto increment.

cc @nickvergessen

lcobucci commented on lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php:
```diff
@@ -534,12 +535,26 @@ public function getAlterTableSQL(TableDiff $diff)
 
-            if ($columnDiff->hasChanged('default') || $columnDiff->hasChanged('type')) {
+            if ($columnDiff->fromColumn) {
+                $oldTypeIsNumeric = $columnDiff->fromColumn->getType()->getName() === Type::INTEGER || $columnDiff->fromColumn->getType()->getName() === Type::BIGINT;
```
Please don't rely on `Type::getName()`, use `instance of` instead.

lcobucci commented on lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php:
```diff
@@ -534,12 +535,26 @@ public function getAlterTableSQL(TableDiff $diff)
-            if ($columnDiff->hasChanged('default') || $columnDiff->hasChanged('type')) {
+            if ($columnDiff->fromColumn) {
+                $oldTypeIsNumeric = $columnDiff->fromColumn->getType()->getName() === Type::INTEGER || $columnDiff->fromColumn->getType()->getName() === Type::BIGINT;
+                $newTypeIsNumeric = $column->getType()->getName() === Type::INTEGER || $column->getType()->getName() === Type::BIGINT;
```
Please don't rely on `Type::getName()`, use `instance of` instead.

lcobucci commented on tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php:
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
         self::assertEquals('(-1)', $columns['col_string']->getDefault());
     }
+
+    public function testAlterTableAutoIncrementBigInt()
```
` : void` missing

lcobucci replied:
Please add `@group 2916` to a docblock here

lcobucci commented on tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php:
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
+
+    public function testAlterTableAutoIncrementBigInt()
+    {
+        $tableFrom = new \Doctrine\DBAL\Schema\Table('autoinc_table_bitgint');
```
Please simplify the FQCN

lcobucci commented on tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php:
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
+        $tableFrom = $this->_sm->listTableDetails('autoinc_table_bitgint');
+        self::assertTrue($tableFrom->getColumn('id')->getAutoincrement());
+
+        $tableTo = new \Doctrine\DBAL\Schema\Table('autoinc_table_bitgint');
```
Please simplify the FQCN

lcobucci commented on tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php:
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
+        $column = $tableTo->addColumn('id', 'bigint');
+        $column->setAutoincrement(true);
+
+        $c = new \Doctrine\DBAL\Schema\Comparator();
```
Please simplify the FQCN

lcobucci commented on tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php:
```diff
@@ -419,6 +419,29 @@ public function testListNegativeColumnDefaultValue()
+        $c = new \Doctrine\DBAL\Schema\Comparator();
+        $diff = $c->diffTable($tableFrom, $tableTo);
+        self::assertInstanceOf('Doctrine\DBAL\Schema\TableDiff', $diff, "There should be a difference and not false being returned from the table comparison");
+        self::assertEquals(array("ALTER TABLE autoinc_table_bitgint ALTER id TYPE BIGINT"), $this->_conn->getDatabasePlatform()->getAlterTableSQL($diff));
```
`assertSame()` is more appropriated here

lcobucci replied:
Please also use short-array syntax

lcobucci reviewed (changes requested):
Please rebase your branch to sync to the latest changes of `master`

icewind1991 committed:
Fix changing column from int to bigint with autoincrement

- SERIAL and BIGSERIAL are not true types and can't be used in ALTER TABLE expressions
- When changing between int and bigint auto increment fields the default should no be dropped
  to preserve the link between the column and the sequence.

icewind1991 commented:
@lcobucci all done

lcobucci commented on tests/Doctrine/Tests/DBAL/Functional/Schema/PostgreSqlSchemaManagerTest.php:
```diff
@@ -463,6 +465,32 @@ public function testAutoIncrementCreatesSerialDataTypesWithoutADefaultValueEvenW
+
+        $c = new Comparator();
+        $diff = $c->diffTable($tableFrom, $tableTo);
+        self::assertInstanceOf('Doctrine\DBAL\Schema\TableDiff', $diff, "There should be a difference and not false being returned from the table comparison");
```
Please use `TableDiff::class` instead

lcobucci commented on lib/Doctrine/DBAL/Platforms/PostgreSqlPlatform.php:
```diff
@@ -536,12 +537,26 @@ public function getAlterTableSQL(TableDiff $diff)
             }
 
-            if ($columnDiff->hasChanged('default') || $columnDiff->hasChanged('type')) {
+            if ($columnDiff->fromColumn) {
```
What do you think about extracting this `if` block to a private method? I think that something like would improve the readability a bit:

```php
private function hasDefaultTypeChanged(ColumnDiff $columnDiff, Column $newColumn) : bool
{
    $defaultTypeChanged = $columnDiff->hasChanged('type');

    if ( ! $columnDiff->fromColumn || ! $defaultTypeChanged) {
        return $defaultTypeChanged;
    }

    $oldTypeIsNumeric = $this->isNumericType($columnDiff->fromColumn->getType());
    $newTypeIsNumeric = $this->isNumericType($newColumn->getType());

    return ! ($oldTypeIsNumeric && $newTypeIsNumeric && $newColumn->getDefault() === null)
}
```

We might improve things there (specially naming wise), but I think it makes things easier and more understandable 😄

lcobucci reviewed (changes requested):
Nice job, some more nitpicks 😜

~Just to understand, is this needed only for PGSQL 10+ compatibility? I mean, I didn't see any issue reporting this before.~
~This question is just to categorise this PR correctly.~

I've already checked, this operation breaks badly on PGSQL 9.x too

lcobucci commented:
One last request, could add some info on the commit msg on why this is needed? BTW thanks for ensuring that the commit history is organised =)

lcobucci commented:
Ahhh I forgot one more thing, can you please add a test for the reverse operation too? bigint -> int

icewind1991 commented:
done

lcobucci commented:
@icewind1991 🚢 and backported to `2.6.x` (with some minor improvements). Thanks for your contribution!

icewind1991 commented:
Any chance of getting things backported to 2.5.x? We're stuck with supporting php 5.6/7.0 for the foreseeable future
